File: opencode/packages/opencode/data/working/robotarm_RRT_star_v01.py
import numpy as np
import matplotlib.pyplot as plt
import math
import os
import time

# --- Parameters ---
L1 = 1.0  # Link 1 length
L2 = 1.0  # Link 2 length
OBSTACLES = [
    (1.0, 1.0, 0.4),
    (-0.5, 1.5, 0.3),
    (0.5, -0.5, 0.3)
] # (x, y, radius)

Q_START = np.array([0.0, 0.0])
Q_GOAL = np.array([np.pi/2, np.pi/4])
GOAL_TOL = 0.1
MAX_ITER = 3000
STEP_SIZE = 0.1
SEARCH_RADIUS = 0.5
GOAL_BIAS = 0.1

OUTPUT_DIR = "data/outputs"
OUTPUT_IMG = os.path.join(OUTPUT_DIR, "robotarm_RRT_star_py_v01_result.png")
OUTPUT_DATA = os.path.join(OUTPUT_DIR, "robotarm_RRT_star_py_v01_data.txt")

# --- Helper Functions ---

def forward_kinematics(q):
    """Returns (xe, ye, x_elbow, y_elbow) given q=[theta1, theta2]"""
    x_elbow = L1 * np.cos(q[0])
    y_elbow = L1 * np.sin(q[0])
    xe = x_elbow + L2 * np.cos(q[0] + q[1])
    ye = y_elbow + L2 * np.sin(q[0] + q[1])
    return xe, ye, x_elbow, y_elbow

def check_line_collision(p1, p2, obstacles):
    """Check if line segment p1-p2 intersects with any circular obstacle"""
    p1 = np.array(p1)
    p2 = np.array(p2)
    v = p2 - p1
    
    for ox, oy, r in obstacles:
        center = np.array([ox, oy])
        w = center - p1
        
        c1 = np.dot(w, v)
        c2 = np.dot(v, v)
        
        if c2 == 0:
            dist = np.linalg.norm(center - p1)
        elif c1 <= 0:
            dist = np.linalg.norm(center - p1)
        elif c2 <= c1:
            dist = np.linalg.norm(center - p2)
        else:
            b = c1 / c2
            pb = p1 + b * v
            dist = np.linalg.norm(center - pb)
            
        if dist <= r:
            return True
    return False

def check_collision(q_start, q_end, obstacles):
    """Check collision for robot arm moving from q_start to q_end"""
    steps = 5
    for s in range(steps + 1):
        alpha = s / steps
        q = q_start * (1 - alpha) + q_end * alpha
        xe, ye, x_elbow, y_elbow = forward_kinematics(q)
        
        # Link 1: (0,0) to (x_elbow, y_elbow)
        if check_line_collision([0, 0], [x_elbow, y_elbow], obstacles):
            return True
        # Link 2: (x_elbow, y_elbow) to (xe, ye)
        if check_line_collision([x_elbow, y_elbow], [xe, ye], obstacles):
            return True
    return False

# --- RRT* Class ---

class Node:
    def __init__(self, coord, parent=None, cost=0.0):
        self.coord = coord
        self.parent = parent
        self.cost = cost

def main():
    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)
        
    print(f"Starting RRT* Optimization (Python)...")
    start_time = time.time()
    
    nodes = [Node(Q_START, None, 0.0)]
    
    for i in range(MAX_ITER):
        # 1. Sample
        if np.random.rand() < GOAL_BIAS:
            q_rand = Q_GOAL
        else:
            q_rand = np.random.uniform(-np.pi, np.pi, 2)
            
        # 2. Nearest
        dists = [np.linalg.norm(n.coord - q_rand) for n in nodes]
        nearest_idx = np.argmin(dists)
        nearest_node = nodes[nearest_idx]
        
        # 3. Steer
        direction = q_rand - nearest_node.coord
        dist = np.linalg.norm(direction)
        
        if dist > STEP_SIZE:
            q_new = nearest_node.coord + (direction / dist) * STEP_SIZE
        else:
            q_new = q_rand
            
        # 4. Collision Check
        if not check_collision(nearest_node.coord, q_new, OBSTACLES):
            
            # 5. Choose Parent (Cost Optimization)
            new_cost = nearest_node.cost + np.linalg.norm(q_new - nearest_node.coord)
            best_node = nearest_node
            min_cost = new_cost
            
            # Find neighbors
            neighbor_idxs = [idx for idx, n in enumerate(nodes) if np.linalg.norm(n.coord - q_new) <= SEARCH_RADIUS]
            
            for idx in neighbor_idxs:
                neighbor = nodes[idx]
                cost_via_neighbor = neighbor.cost + np.linalg.norm(q_new - neighbor.coord)
                
                if cost_via_neighbor < min_cost:
                    if not check_collision(neighbor.coord, q_new, OBSTACLES):
                        min_cost = cost_via_neighbor
                        best_node = neighbor
                        
            # Add Node
            new_node = Node(q_new, best_node, min_cost)
            nodes.append(new_node)
            
            # 6. Rewire
            for idx in neighbor_idxs:
                neighbor = nodes[idx]
                cost_via_new = new_node.cost + np.linalg.norm(new_node.coord - neighbor.coord)
                
                if cost_via_new < neighbor.cost:
                    if not check_collision(new_node.coord, neighbor.coord, OBSTACLES):
                        neighbor.parent = new_node
                        neighbor.cost = cost_via_new
                        
        if (i+1) % 500 == 0:
            print(f"Iteration {i+1}/{MAX_ITER}...")

    elapsed_time = time.time() - start_time
    
    # --- Extract Path ---
    dists_to_goal = [np.linalg.norm(n.coord - Q_GOAL) for n in nodes]
    best_goal_idx = np.argmin(dists_to_goal)
    best_goal_node = nodes[best_goal_idx]
    
    if dists_to_goal[best_goal_idx] <= GOAL_TOL:
        print(f"Goal Reached! Cost: {best_goal_node.cost:.4f}")
        path = []
        curr = best_goal_node
        while curr is not None:
            path.append(curr.coord)
            curr = curr.parent
        path = np.array(path[::-1]) # Reverse to start->goal
        success = True
    else:
        print("Failed to reach goal.")
        path = None
        success = False

    # --- Visualization ---
    fig, ax = plt.subplots(1, 2, figsize=(12, 6))
    
    # 1. Workspace
    ax[0].set_title(f"Workspace (Cost: {best_goal_node.cost:.2f})")
    ax[0].set_xlim(-2.5, 2.5)
    ax[0].set_ylim(-2.5, 2.5)
    ax[0].set_aspect('equal')
    ax[0].grid(True)
    
    # Obstacles
    for ox, oy, r in OBSTACLES:
        circle = plt.Circle((ox, oy), r, color='gray')
        ax[0].add_patch(circle)
        
    # Path
    if success:
        for q in path[::max(1, len(path)//10)]: # Sample frames
            xe, ye, x_elbow, y_elbow = forward_kinematics(q)
            ax[0].plot([0, x_elbow, xe], [0, y_elbow, ye], 'b-', alpha=0.2)
            
        # Start & Goal
        xe, ye, x_elbow, y_elbow = forward_kinematics(Q_START)
        ax[0].plot([0, x_elbow, xe], [0, y_elbow, ye], 'g-o', linewidth=2, label='Start')
        xe, ye, x_elbow, y_elbow = forward_kinematics(Q_GOAL)
        ax[0].plot([0, x_elbow, xe], [0, y_elbow, ye], 'r-o', linewidth=2, label='Goal')
        ax[0].legend()
    
    # 2. C-Space
    ax[1].set_title("C-Space (Joint Angles)")
    ax[1].set_xlim(-np.pi, np.pi)
    ax[1].set_ylim(-np.pi, np.pi)
    ax[1].set_aspect('equal')
    ax[1].grid(True)
    ax[1].set_xlabel("Theta 1")
    ax[1].set_ylabel("Theta 2")
    
    # Tree (Sampling)
    # for n in nodes:
    #     if n.parent:
    #         ax[1].plot([n.parent.coord[0], n.coord[0]], [n.parent.coord[1], n.coord[1]], 'k-', alpha=0.1, linewidth=0.5)
            
    if success:
        ax[1].plot(path[:,0], path[:,1], 'r-', linewidth=2)
    
    ax[1].plot(Q_START[0], Q_START[1], 'go')
    ax[1].plot(Q_GOAL[0], Q_GOAL[1], 'ro')
    
    plt.savefig(OUTPUT_IMG)
    print(f"Saved plot to {OUTPUT_IMG}")
    
    # --- Save Data ---
    with open(OUTPUT_DATA, "w") as f:
        f.write(f"Success: {success}\n")
        f.write(f"Cost: {best_goal_node.cost if success else 'inf'}\n")
        f.write(f"Time: {elapsed_time:.4f}\n")
        f.write(f"Iterations: {MAX_ITER}\n")
        f.write(f"Nodes: {len(nodes)}\n")
        if success:
            f.write("Path:\n")
            np.savetxt(f, path)
            
    print(f"Saved data to {OUTPUT_DATA}")

if __name__ == "__main__":
    main()

File: opencode/packages/opencode/data/matlab/robotarm_RRT_star_v01.m
% robotarm_RRT_star_v01.m
% 2-Link Robot Arm Path Planning using RRT* (Optimal RRT)
% Features:
% - 2-Link Planar Arm Kinematics
% - Configuration Space (C-Space) Sampling
% - Link-level Collision Detection (not just end-effector)
% - Path Optimization (Rewiring)

clear; clc; close all;

%% 1. Experiment Setup
% Robot Parameters
L1 = 1.0; % Length of link 1
L2 = 1.0; % Length of link 2
link_width = 0.1;

% Environment (Obstacles: [x, y, radius])
obstacles = [
    1.0, 1.0, 0.4;
    -0.5, 1.5, 0.3;
    0.5, -0.5, 0.3
];

% Task Definition (Joint Angles in Radians)
q_start = [0, 0];          % Start configuration
q_goal = [pi/2, pi/4];     % Goal configuration
goal_tol = 0.1;            % Tolerance to reach goal

% RRT* Parameters
max_iter = 3000;
step_size = 0.1;           % Max angle change per step
search_radius = 0.5;       % Rewiring radius
goal_bias = 0.1;           % Probability to sample goal

%% 2. Initialization
nodes(1).coord = q_start;  % Joint angles [theta1, theta2]
nodes(1).parent = 0;
nodes(1).cost = 0;
nodes(1).coord_end = forward_kinematics(q_start, L1, L2); % End effector pos (for vis)

% Visualization Setup
f = figure('Color', 'w', 'Position', [100, 100, 1000, 500]);
subplot(1,2,1); hold on; grid on; axis equal;
title('Workspace (Robot Arm)');
xlabel('X (m)'); ylabel('Y (m)');
axis([-2.5 2.5 -2.5 2.5]);

% Draw Obstacles
for i = 1:size(obstacles, 1)
    rectangle('Position', [obstacles(i,1)-obstacles(i,3), obstacles(i,2)-obstacles(i,3), ...
               2*obstacles(i,3), 2*obstacles(i,3)], 'Curvature', [1 1], 'FaceColor', [0.2 0.2 0.2]);
end

subplot(1,2,2); hold on; grid on; axis equal;
title('C-Space (Joint Angles)');
xlabel('\theta_1 (rad)'); ylabel('\theta_2 (rad)');
axis([-pi pi -pi pi]);
plot(q_start(1), q_start(2), 'go', 'MarkerFaceColor', 'g');
plot(q_goal(1), q_goal(2), 'ro', 'MarkerFaceColor', 'r');

%% 3. RRT* Main Loop
fprintf('Starting RRT* Optimization...\n');

for i = 1:max_iter
    % 3.1 Sampling
    if rand < goal_bias
        q_rand = q_goal;
    else
        q_rand = [rand*2*pi - pi, rand*2*pi - pi]; % Random [-pi, pi]
    end
    
    % 3.2 Nearest Node
    dists = arrayfun(@(n) norm(n.coord - q_rand), nodes);
    [~, nearest_idx] = min(dists);
    q_near = nodes(nearest_idx).coord;
    
    % 3.3 Steer
    dir = q_rand - q_near;
    dist = norm(dir);
    if dist > step_size
        q_new = q_near + (dir / dist) * step_size;
    else
        q_new = q_rand;
    end
    
    % 3.4 Collision Check (Robot Body)
    if ~check_collision(q_near, q_new, L1, L2, obstacles)
        
        % 3.5 Choose Parent (Cost Optimization)
        min_cost = nodes(nearest_idx).cost + norm(q_new - q_near);
        best_parent = nearest_idx;
        
        % Find neighbors within radius
        neighbor_idxs = find(dists <= search_radius);
        
        for idx = neighbor_idxs
            cost_via_neighbor = nodes(idx).cost + norm(q_new - nodes(idx).coord);
            if cost_via_neighbor < min_cost
                if ~check_collision(nodes(idx).coord, q_new, L1, L2, obstacles)
                    min_cost = cost_via_neighbor;
                    best_parent = idx;
                end
            end
        end
        
        % Add Node
        new_idx = length(nodes) + 1;
        nodes(new_idx).coord = q_new;
        nodes(new_idx).parent = best_parent;
        nodes(new_idx).cost = min_cost;
        
        % 3.6 Rewire (Optimize Neighbors)
        for idx = neighbor_idxs
            cost_via_new = nodes(new_idx).cost + norm(nodes(idx).coord - q_new);
            if cost_via_new < nodes(idx).cost
                if ~check_collision(q_new, nodes(idx).coord, L1, L2, obstacles)
                    nodes(idx).parent = new_idx;
                    nodes(idx).cost = cost_via_new;
                    % Note: Strictly, we should update children costs recursively here,
                    % but skipped for basic performance.
                end
            end
        end
        
        % Plotting (C-Space)
        if mod(i, 50) == 0
            subplot(1,2,2);
            plot([nodes(best_parent).coord(1), q_new(1)], [nodes(best_parent).coord(2), q_new(2)], 'b-', 'Color', [0.8 0.8 0.8]);
            plot(q_new(1), q_new(2), 'b.', 'MarkerSize', 2);
            drawnow limitrate;
        end
        
        % Check Goal
        if norm(q_new - q_goal) < goal_tol
            fprintf('Goal region reached at iter %d! Continuing for optimization...\n', i);
        end
    end
end

%% 4. Extract Path
% Find node closest to goal
dists_to_goal = arrayfun(@(n) norm(n.coord - q_goal), nodes);
[min_dist_goal, goal_idx] = min(dists_to_goal);

if min_dist_goal < goal_tol
    fprintf('Path Found! Cost: %.4f\n', nodes(goal_idx).cost);
    
    % Backtrack
    curr = goal_idx;
    path_idx = [];
    while curr ~= 0
        path_idx = [curr, path_idx];
        curr = nodes(curr).parent;
    end
    
    path_q = reshape([nodes(path_idx).coord], 2, [])';
    
    % Visualize Final Path
    subplot(1,2,2);
    plot(path_q(:,1), path_q(:,2), 'r-', 'LineWidth', 2);
    
    subplot(1,2,1);
    for j = 1:size(path_q, 1)
        q = path_q(j, :);
        [xe, ye, x_elbow, y_elbow] = forward_kinematics(q, L1, L2);
        
        % Draw Robot
        % Clear previous robot lines (tricky in loop, just draw over or use handle)
        if j > 1; delete(h_arm); end
        h_arm = plot([0, x_elbow, xe], [0, y_elbow, ye], 'b-o', 'LineWidth', 2, 'MarkerFaceColor', 'b');
        pause(0.05);
    end
    title('Robot Arm Motion (Optimized)');
else
    fprintf('Failed to reach goal within tolerance.\n');
end

%% Helper Functions

function [xe, ye, x_elbow, y_elbow] = forward_kinematics(q, L1, L2)
    x_elbow = L1 * cos(q(1));
    y_elbow = L1 * sin(q(1));
    xe = x_elbow + L2 * cos(q(1) + q(2));
    ye = y_elbow + L2 * sin(q(1) + q(2));
end

function collision = check_collision(q_start, q_end, L1, L2, obstacles)
    % Interpolate between configurations to check swept volume
    steps = 10;
    collision = false;
    
    for s = 0:steps
        alpha = s / steps;
        q = q_start * (1 - alpha) + q_end * alpha;
        [xe, ye, x_elbow, y_elbow] = forward_kinematics(q, L1, L2);
        
        % Check Link 1 (Origin to Elbow)
        if check_line_collision([0, 0], [x_elbow, y_elbow], obstacles)
            collision = true; return;
        end
        
        % Check Link 2 (Elbow to End-Effector)
        if check_line_collision([x_elbow, y_elbow], [xe, ye], obstacles)
            collision = true; return;
        end
    end
end

function hit = check_line_collision(p1, p2, obstacles)
    hit = false;
    for i = 1:size(obstacles, 1)
        ox = obstacles(i,1); oy = obstacles(i,2); r = obstacles(i,3);
        % Distance from point (center of obstacle) to line segment p1-p2
        % Vector formulation
        v = p2 - p1;
        w = [ox, oy] - p1;
        
        c1 = dot(w, v);
        c2 = dot(v, v);
        
        if c2 == 0 % p1 and p2 are same
            dist = norm([ox, oy] - p1);
        elseif c1 <= 0
            dist = norm([ox, oy] - p1);
        elseif c2 <= c1
            dist = norm([ox, oy] - p2);
        else
            b = c1 / c2;
            pb = p1 + b * v;
            dist = norm([ox, oy] - pb);
        end
        
        if dist <= r
            hit = true; return;
        end
    end
end
